src/api/MainSolver.h

```diff
@@ -236,6 +238,9 @@ class MainSolver {
     // Returns model of the last query (must be in satisfiable state)
     std::unique_ptr<Model> getModel();
 
+    // Returns proof of the last query (must be in unsatisfiable state)
+    ResolutionProof const & getResolutionProof() const;
```

Comment:
I am actually starting to question if this should be exposed.

What if  `MainSolver` would only have a method to print the proof, but would never actually expose `ResolutionProof` to the outside world?

If some program includes OpenSMT as a library it will see this method on `MainSolver` object, but it will not be able to do anything with the reference, because `ResolutionProof` will be incomplete type at that point, no?

Reply:
A lot of classes in OpenSMT are designed as separate classes, not as nested classes. Hence it is problematic to "hide" them from the users. I honestly do not like too much the design where it is necessary to forward-declare classes and to define them later, I think that the header should be always available. In the case of inappropriate access to a "hidden" class, I think it is much cleaner and clearer to let the compiler complain that the class is protected/private than to have undefined references, where it is very unclear why it happened.

Anyway, I do not understand the class `ResolutionProof` that much so I cannot say if there are use cases where the user indeed wants the class itself. If not, I agree to remove it. However, this would not remove the problem, `ResolutionProof` is still available from `CoreSMTSolver` which is accessible from `MainSolver`.

Reply:
`CoreSMTSolver` should also be hidden from the users.
My mental model is that `MainSolver` should work as the API of OpenSMT, exposing all the functionality without exposing the implementation. I see `CoreSMTSolver` and `ResolutionProof` as the implementation details.
Maybe there is a case for exposing some version of a proof to the user, but at the moment I don't believe our `ResolutionProof` is the abstraction that should be exposed.